subprojects/core/src/main/java/org/gradle/execution/ProjectExecutionServices.java

```diff
@@ -137,14 +137,14 @@ TaskExecuter createTaskExecuter(
         WorkExecutor<ExecutionRequestContext, CachingResult> workExecutor
     ) {
 
-        ExecuteActionsTaskExecuter.VfsInvalidationStrategy vfsInvalidationStrategy = VirtualFileSystemServices.isPartialInvalidationEnabled(startParameter.getSystemPropertiesArgs())
+        ExecuteActionsTaskExecuter.VfsInvalidationStrategy vfsInvalidationStrategy = VirtualFileSystemServices.isPartialInvalidationEnabled(startParameter)
             ? ExecuteActionsTaskExecuter.VfsInvalidationStrategy.PARTIAL
             : ExecuteActionsTaskExecuter.VfsInvalidationStrategy.COMPLETE;
 
         // TODO: The incubation message should be printed in VirtualFileSystemServices.
         //   The problem is that `RootBuildLifecycleListener.afterStart` is called to early to have the system properties from gradle.properties available
         //   We log the message now here as a workaround.
-        if (vfsInvalidationStrategy == ExecuteActionsTaskExecuter.VfsInvalidationStrategy.PARTIAL && !VirtualFileSystemServices.isRetentionEnabled(startParameter.getSystemPropertiesArgs())) {
+        if (vfsInvalidationStrategy == ExecuteActionsTaskExecuter.VfsInvalidationStrategy.PARTIAL && !startParameter.isWatchFileSystem()) {
```

Comment:
Would it be possible to inject `WatchFileSystemOption` here instead?

Reply:
I don't think build options can be injected.

Reply:
What I think I'm trying to say is that they should be. That way we could precisely inject things that matter to the places where they are needed instead of dragging the whole mutable mess that is `StartParameter` currently.

Reply:
👍